### PR TITLE
storage/engine: fix prefix iteration in BaseDeltaIterator

### PR DESCRIPTION
**Pull request by petermattis**
This change is [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/6996)

### CODE REVIEW

File: storage/engine/rocksdb/db.cc [hunk 1]
```diff
@@ -1105,32 +1105,6 @@ class BaseDeltaIterator : public rocksdb::Iterator {
...
-    // current_at_base -> compare < 0
-    assert(!current_at_base_ || compare < 0);
-    // !current_at_base -> compare <= 0
-    assert(current_at_base_ && compare >= 0);
-    // equal_keys_ <=> compare == 0
-    assert((equal_keys_ || compare != 0) && (!equal_keys_ || compare == 0));
-#endif
-  }
-
   void Advance() {
```
Comment (tschottdorf) [inline storage/engine/rocksdb/db.cc]: Could you add comments on all of the methods here? CheckPrefix, ProcessDelta, ... - I was reading all of this code and it wasn't fun to try to guess what it wants to do. Ditto for the base and delta iterator and the other miscellaneous state on this object.

**Comment by tschottdorf** [comment]
Essentially LGTM, but would really appreciate more commentary on the file as a whole (especially on base vs. delta iterator).

**Comment by petermattis** [comment]
Review status: 0 of 2 files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

*[storage/engine/rocksdb/db.cc, line 1108 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6996#-KJCn5vPBDkW4Dzx5I37-r2-1108:-KJDUVJDKUB0h4Uam49v:1869608525) ([raw file](https://github.com/cockroachdb/cockroach/blob/e1ea9f6d57b0e821dff1ee57af49696e8329e439/storage/engine/rocksdb/db.cc#L1108)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> Could you add comments on all of the methods here? CheckPrefix, ProcessDelta, ... - I was reading all of this code and it wasn't fun to try to guess what it wants to do. Ditto for the base and delta iterator and the other miscellaneous state on this object.
</details>
Added a bunch of commentary on `BaseDeltaIterator`. Sort of irritating because most of this code is inherited from upstream and is uncommented there. PTAL.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6996)*

File: storage/engine/rocksdb/db.cc [hunk 2]
```diff
@@ -1237,19 +1236,19 @@ class BaseDeltaIterator : public rocksdb::Iterator {
...
+        // Delta is greater than base (use base).
         current_at_base_ = true;
         return;
       }
-      // Delta is less than or equal to base.
-      if (check_prefix && CheckPrefix(delta_iterator_->Entry().key)) {
-        return;
-      }
+      // Delta is less than or equal to base. If check_prefix is true,
+      // for base to be valid is has to contain the prefix we were
```
Comment (tschottdorf) [inline storage/engine/rocksdb/db.cc]: s/is/it/

**Comment by petermattis** [comment]
Merging in order to unblock @tamird. I'll send a follow-up PR if there is additional commentary.

File: storage/engine/rocksdb/db.cc [hunk 3]
```diff
@@ -1278,15 +1276,27 @@ class BaseDeltaIterator : public rocksdb::Iterator {
...
   mutable rocksdb::Status status_;
+  // The merged delta value returned when we're pointed at the delta
+  // iterator.
   mutable DBString merged_;
+  // The base iterator, presumably obtained from a rocksdb::DB.
   std::unique_ptr<rocksdb::Iterator> base_iterator_;
+  // The delta iterator obtained from a rocksdb::WriteBatchWithIndex.
   std::unique_ptr<rocksdb::WBWIIterator> delta_iterator_;
+  // The key the delta iterator is currently pointed at. We can't use
+  // delta_iterator_->Entry().key due to the handling merge
```
Comment (tschottdorf) [inline storage/engine/rocksdb/db.cc]: s/the//?

File: storage/engine/rocksdb/db.cc [hunk 4]
```diff
@@ -1179,10 +1160,14 @@ class BaseDeltaIterator : public rocksdb::Iterator {
     return merged_.data == NULL;
```
Comment (tschottdorf) [inline storage/engine/rocksdb/db.cc]: What's going on in l. 1154 to here? Why have we advanced past the last entry for key?

**Comment by tschottdorf** [comment]
Much of the code still looks like it's not getting much exercise (for example `ProcessDeltaKey`). Considering that we've on more than one occasion spent a lot of time into debugging and fixing the bugs that do crop up. Do you feel confident about that part of the codebase? Improving coverage here should be filed as an issue (could you do it? I'm not sure what the best way to exercise more C++ is).

File: storage/engine/rocksdb/db.cc (hunk 4 above)
Comment (petermattis) [inline storage/engine/rocksdb/db.cc]: `ProcessDeltaKey` advances the iterator until either the delta iterator is not valid or the key has changed. In either case, we've advanced past the last entry for the current key.

File: storage/engine/rocksdb/db.cc (hunk 2 above)
Comment (petermattis) [inline storage/engine/rocksdb/db.cc]: Will send a follow-on PR.

**Comment by tschottdorf** [comment]
Thanks for the commentary. Much much better. It's a shame there isn't more of it upstream already.

File: storage/engine/rocksdb/db.cc (hunk 3 above)
Comment (petermattis) [inline storage/engine/rocksdb/db.cc]: `due to the handling of merge`

File: storage/engine/rocksdb/db.cc (hunk 4 above)
Comment (tschottdorf) [inline storage/engine/rocksdb/db.cc]: Ah, gotcha. Would be useful in the comment on `ProcessDeltaKey`.

**Comment by petermattis** [comment]
`ProcessDeltaKey` is exercised by various tests in `batch_test.go` (e.g. `TestBatchMerge`). I'm not aware of any lack of coverage in this code. What we saw here was clearly a bug and a hole in coverage, but that doesn't mean the rest of the code isn't being tested. FWIW, I don't think this code is our weakest link in terms of stability. And if I had concrete knowledge of where coverage was lacking in this code I would fix it (as I've done in this PR).

I agree that debugging this code is certainly irritating (and we've done so on more than one occasion), but those occasions are rare in comparison to bugs in other parts of the system. The big irritation is that you question your sanity when there are bugs in this layer because we've been conditioned to expect it to be solid.

File: storage/engine/rocksdb/db.cc (hunk 4 above)
Comment (petermattis) [inline storage/engine/rocksdb/db.cc]: Added a comment in the follow-on PR.

**Pull request merged**